source/extensions/filters/network/ratelimit/ratelimit.h

```diff
@@ -83,13 +87,47 @@ class Filter : public Network::ReadFilter,
   void initializeReadFilterCallbacks(Network::ReadFilterCallbacks& callbacks) override {
     filter_callbacks_ = &callbacks;
     filter_callbacks_->connection().addConnectionCallbacks(*this);
+    applySubstitutionFormatter(config_->descriptors(),
+                               filter_callbacks_->connection().streamInfo());
+  }
+
+  void applySubstitutionFormatter(std::vector<RateLimit::Descriptor> original_descriptors,
+                                  StreamInfo::StreamInfo& stream_info) {
+
+    std::vector<RateLimit::Descriptor> dynamicDescriptors = std::vector<RateLimit::Descriptor>();
+    for (const RateLimit::Descriptor& descriptor : original_descriptors) {
+      RateLimit::Descriptor new_descriptor;
+      for (const RateLimit::DescriptorEntry& descriptorEntry : descriptor.entries_) {
+        std::string value = descriptorEntry.value_;
+        if (config_->runtime().snapshot().runtimeFeatureEnabled(
+                "envoy.reloadable_features.network.rate_limit.dynamic_downstream_ip") &&
+            absl::StrContains(value, "envoy.downstream_ip")) {
+          value =
+              substitutionFormattedString("%DOWNSTREAM_REMOTE_ADDRESS_WITHOUT_PORT%", stream_info);
```

Comment:
The way this should work is the entire value should be fed into a substitution formatter, much like we do for access logging. Any value that the formatter supports should work, as long as stream info is populated. Essentially, almost any value that tcp_proxy supports should work here. This will make it much more general purpose.

Reply:
I've changed the implementation to a substitution formatter based approach. Can you take a look and see if this is what you envisioned? And what sort of documentation you'd like to see?